pkg/beholder/client_test.go

```diff
@@ -0,0 +1,353 @@
+		},
+		{
+			name:                "Valid URI",
+			exporterCalledTimes: 1,
+			attrs: Attributes{
+				"beholder_data_schema": "/example-schema/versions/1",
+			},
+			expectedError: "",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			setupMock := func(exporterMock *mocks.OTLPExporter) (*mocks.OTLPExporter, <-chan struct{}) {
+				done := make(chan struct{}, tc.exporterCalledTimes)
+				if tc.exporterCalledTimes > 0 {
+					exporterMock.On("Export", mock.Anything, mock.Anything).Return(nil).Times(tc.exporterCalledTimes).
+						Run(func(args mock.Arguments) {
+							done <- struct{}{}
+						})
+				}
+				return exporterMock, done
+			}
+
+			assertError := func(err error, expected string) {
+				if tc.expectedError != "" {
+					assert.ErrorContains(t, err, expected)
+				} else {
+					assert.NoError(t, err)
+				}
+			}
+
+			assertMock := func(exporterMock *mocks.OTLPExporter) {
+				if tc.exporterCalledTimes > 0 {
+					exporterMock.AssertExpectations(t)
+				} else {
+					exporterMock.AssertNotCalled(t, "Export")
+				}
+			}
+
+			waitUntilSent := func(done <-chan struct{}) {
+				for range tc.exporterCalledTimes {
+					select {
+					case <-done:
+					case <-time.After(10 * time.Second):
+						t.Fatalf("Timed out waiting for messages to be emitted")
+					}
+				}
+			}
+
+			setupTest := func() (emitter Emitter, message Message, assertExpectations func(err error)) {
+				exporterMock, done := setupMock(mocks.NewOTLPExporter(t))
+				emitter = getEmitter(exporterMock)
+				message = NewMessage([]byte("test"), tc.attrs)
+
+				assertExpectations = func(err error) {
+					assertError(err, tc.expectedError)
+					if err == nil {
+						waitUntilSent(done)
+					}
+					assertMock(exporterMock)
+				}
+				return
+			}
+			t.Run("Emitter.Emit", func(t *testing.T) {
+				emitter, message, assertExpectations := setupTest()
+
+				err := emitter.Emit(tests.Context(t), message.Body, tc.attrs)
+
+				assertExpectations(err)
+			})
+		})
+	}
+}
+
+func TestClient_Close(t *testing.T) {
+	exporterMock := mocks.NewOTLPExporter(t)
+	defer exporterMock.AssertExpectations(t)
+
+	client := NewStdoutClient()
+
+	err := client.Close()
+	assert.NoError(t, err)
+
+	exporterMock.AssertExpectations(t)
+}
+
+func TestClient_SetGlobas(t *testing.T) {
+	exporterMock := mocks.NewOTLPExporter(t)
+	defer exporterMock.AssertExpectations(t)
+
+	otelErrorHandler := func(err error) {
+		t.Fatalf("otel error: %v", err)
+	}
+	// Override exporter factory which is used by Client
+	exporterFactory := func(context.Context, ...otlploggrpc.Option) (sdklog.Exporter, error) {
+		return exporterMock, nil
+	}
+	client, err := newOtelClient(DefaultConfig(), otelErrorHandler, exporterFactory)
+	if err != nil {
+		t.Fatalf("Error creating beholder client: %v", err)
+	}
+	assert.NoError(t, err)
+
+	globals := getGlobals()
+	defer restoreGlobas(t, globals)
+
+	client.SetGlobals()
```

Comment:
This feels odd to me. The flow I would expect is that you set your global providers in some type of init, and then inherit those providers in your clients. The globals are "read-only" with respect to the later initialized clients that need to configure those providers.

thoughts @jmank88 ?

Reply:
I think it depends how we want to coordinate the resource, but it does seems more composable to separate the concern of initing the global provider and registering the exporters etc.

Reply:
@patrickhuie19 @jmank88 see https://github.com/smartcontractkit/chainlink-common/pull/694/commits/faa61c20702f1c1ccc1bebd6ca2e64c450f63df1